### ISSUE TITLE
Missing some basic features

### ISSUE THREAD

**Issue by hidegh** [VeritasSoftware/ts.validator#3]
1. The code below throws an exception, as there's no **NotNullNorEmpty** rule
2. The **identifier is not auto filled** based on the fat-arrow property selector - **seems this is just an issue when nesting**
3. For **basic rules** (required, min/max, length, etc.) a **default messages**: {0} is required (where {0} is the identifier, {1} or {2} or {3} are additional parameters in case of min-max length would be welcome - this way messages could be altered by user more easily (also localisation would be possible later on)
4. Also miss **rule and/or sample for nesting** (similar to the If rule) - i faked it with the NotNull HomeAddress, but basically a rule should be used that if a nested object has some rules, then those rules should be triggered if nested object is undefined | null. So maybe **the nesting rule should have an optional condition as well.**

```
export class AppComponent {
  name = "Angular";
  data: any;
  result: any;

  constructor() {
    this.data = PersonModel.getSample();

    let rules = (validator: IValidator<PersonModel>): ValidationResult => {
      return validator
        .IsEmpty(m => m.FirstName, "First name must be empty")
        .NotNull(m => m.HomeAddress, "Home address is mandatory")
        .If(m => m.HomeAddress, validator => validator
          //.NotEmpty(m => m.HomeAddress.Street, "Street is mandatory")
          .NotNull(m => m.HomeAddress.Street, "Street is mandatory")
          .ToResult()
        ) 
        .ToResult();
    };

    this.result = new Validator(this.data).Validate(rules);   

  }
}
```

https://stackblitz.com/edit/angular-ts-flunt-validator?file=src%2Fapp%2Fapp.component.ts

**Comment by VeritasSoftware** [comment]
Thanks. The framework is pretty extensive as it is.

I will see if any in your wish list can be implemented.
I will go though your stackblitz code.

**Comment by hidegh** [comment]
@VeritasSoftware
1. Here's a sample on **getting the full path** from the expression, shoulf fix issue No. 2. (see code below) - **just replace** the objectValidator.ts::getPropertyName function code
2. The **NotNullOrEmpty** is a common case (see .NET IsNullOrWhitespace, IsNullOrEmpty)
3. And **nested** components are common, if nested component is required then all inner rules should be executed (even is nested value is null, so in that case all properties we should consider as null).

In the past I found this library, which I patched in 2 places:
1. the property name extraction (especially for nested property selection via fat arrow)
2. auto setting the param (so the property name) in the output
https://github.com/coditorium/nodejs-fluent-validator/issues/4

```
  memberNamesExtractor = new RegExp("return (.*);?\\b");
  memberNamesExtractorArrowFunctions = new RegExp("=>(.*)");

  public getPropertySelector<TResult>(name: (x?: TResult) => any) {
    let match =
      this.memberNamesExtractor.exec(name + "") ||
      this.memberNamesExtractorArrowFunctions.exec(name + "");

    if (match == null)
      throw new Error(
        "The function does not contain a statement matching 'return variableName;'"
      );

    let expression = match[1].toString();

    let firstDotIndex = expression.indexOf(".");
    let propertySelector = expression.substring(firstDotIndex + 1);

    return propertySelector;
  }
```
